### PR TITLE
Sphinx cleanups, bump version requirements, typecheck docs source files

### PR DESCRIPTION
**Pull request by jakkdl**
I was struggling with sphinx in #2898 and noticed that sphinx was pinned to an ancient version, and unpinning & updating it seemed to fix a couple of them. My struggles also made me understand sphinx innards better than I had before, so could fix a bunch of references at the same time.

* add missing type annotations to docs/source/*.py
* specify files to check with mypy in pyproject.toml (docs/source/typevars.py was typed already, but not checked in CI)
* change sphinx version limits to >5.3
* clean up & fix a bunch of references in sphinx
* add comment explaining workarounds

let's see if RTD manages this as well

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -41,6 +49,20 @@
             check=True,
         )
 
+# Sphinx is very finicky, and somewhat buggy, so we have several different
+# methods to help it resolve links.
```
Comment (TeamSpen210) [inline docs/source/conf.py:53]: Might want to mention the `typevar` module also, which redirects `typing_extensions` → `typing`, and all typevars to `typing.TypeVar` etc.

File: docs/source/conf.py [hunk 2]
```diff
@@ -16,8 +16,16 @@
...
+from __future__ import annotations
+
+import collections.abc
 import os
 import sys
+from typing import TYPE_CHECKING, cast
+
+if TYPE_CHECKING:
+    from sphinx.application import Sphinx
+    from sphinx.util.typing import Inventory
```
Comment (TeamSpen210) [inline docs/source/conf.py:28]: Not sure if the guard is necessary. If this is running, Sphinx is installed?

Comment (Fuyukai) [reply docs/source/conf.py:28]: Ruff does this automatically for any objects that are only found in type hints for some reason

Comment (TeamSpen210) [reply docs/source/conf.py:28]: Ahh right.

Comment (jakkdl) [reply docs/source/conf.py:28]: Hah, I did it manually out of habit. I like denoting typing-only imports

**Review by TeamSpen210 — APPROVED** [review]

**Pull request merged**